libraries/chain/controller.cpp

```diff
@@ -317,17 +318,24 @@ struct controller_impl {
             ("s", start_block_num)("n", blog_head->block_num()) );
 
       auto start = fc::time_point::now();
+      bool cerr_is_tty = isatty(fileno_hack(std::cerr));
```

Comment:
I'm a little confused by the need for fileno_hack here. stderr is, by definition, always fd 2. Why can't this code just be `bool cerr_is_tty = isatty(STDERR_FILENO);` and call it a day.

Reply:
The constant stderr is indeed always fd 2, but the actual stream may no longer be fd 2 if it has been manipulated.  Specifically, in some versions of macOS, `fclose()` of stdin or stdout followed by an `freopen()` of stderr will change the file descriptor to the first available one, which could be 0 or 1.  This behavior is strictly POSIX compliant because of the wording of the description of `freopen()`, which reads "The original stream shall be closed regardless of whether the **subsequent** open succeeds."  [emphasis added]  Various standard libraries, including glibc, are intentionally noncompliant, but macOS strictly complied in at least 10.9.4 and possibly other versions.

Reply:
okay, fair enough, but that's something our process would have to explicitly do. And we know it doesn't do that. So is it worth dipping our hand in to the private implementation of the c++ library for something we'll never do?

Reply:
Per opinion of attorneys, CC-BY-SA licensed code can not be used without changing the current license.